### PR TITLE
Performance changes

### PR DESCRIPTION
**Pull request by nicholasklogan**
Hello Again,

This is the performance notes I made: again I recommend going through the [commits ](https://github.com/AlexandreSajus/Python-Ecosystem/pull/2/commits)to explain what was changed and why. I forgot to mention it before but you can click on lines and leave comments/questions/etc. if you want.

To perform the performance (pun not intended) improvements I ran:

`python -m cProfile -s tottime ./run.py > FILENAME`
This runs your normal program and produces what I call a benchmark, it is by no means perfect but gives you a starting point to work from.

I then create tests to run the worst offenders and see if any if the sub-functions are to blame and to give a consistent run to compare improvements against. I also tell it to run the test thousands or millions of times to get a better average of run times. This is run with something like this: `python -m cProfile -s cumtime ./tests/test_act.py`.

If the code is complex or it is hard to identify here the issue in the code is. I recommend breaking the code into sub-functions (even if temporarily) to see which chunk of code/sub-function is the slowest. as seen in this [commit](https://github.com/AlexandreSajus/Python-Ecosystem/commit/6474f0fe5135be6892fc4f0850d5d475d63df9ee); A side benefit of this is that it makes the code a bit easier to read.

For more info on profiling code in python see [here](https://docs.python.org/3/library/profile.html#the-python-profilers)

In simple terms the:

-  first column of the profiler output is times the function as called
-  second column is time spent in the function running the direct code of that function
-  third is time in that direct function per call
-  fourth is time spent in that function and all sub-functions
-  fifth is time spent in that function and all sub-functions per call
-  sixth is the file, function, and line where that function starts

Note: the reason I didn't use the `-o` flag on the cProfiler is because it writes in a strange format that is not useful here, but is useful for some things (like using `py2proftree` and `kcachegrind` to visualize the code) <- check this out if you are interested I really think its pretty cool.

### CODE REVIEW

**Comment by AlexandreSajus** [comment]
Wow again! Thank you very much for teaching me all this.

The only benchmarking I ever did was with timeit but this is way more efficient to find performance improvements. How did you find out about cProfile? I have never heard of it. Where can I find out more about similar tools that I might not know about?

**Comment by nicholasklogan** [comment]
> Wow again! Thank you very much for teaching me all this.
> 
> The only benchmarking I ever did was with timeit but this is way more efficient to find performance improvements. How did you find out about cProfile? I have never heard of it. Where can I find out more about similar tools that I might not know about?

cProfile is something I picked up years ago, its builtin to python. Not sure exactly when I discovered it. I read the entire [Programming Python](https://www.oreilly.com/library/view/programming-python-4th/9781449398712/) book a few times years ago along with a few like [High Performance Python](https://www.oreilly.com/library/view/high-performance-python/9781492055013/).

There are a lot of profiling, coverage, etc. tools out there. I would honestly google it. I can recommend: a series of[ youtube videos](https://youtube.com/playlist?list=PL7d-X6ouyM17SaXqtW7gLfeFBPOInnues) that I like that discuss good coding practices, also a tool I like called [vulture](https://pypi.org/project/vulture/) for finding dead code. Also, I recommend looking into [Test Driven Development](https://www.agilealliance.org/glossary/tdd/) (TDD) and [London-Style Development](https://softwareengineering.stackexchange.com/questions/123627/what-are-the-london-and-chicago-schools-of-tdd) to help with structuring your code.
